service/qos/service_level_controller: update cache on startup

mlitvk opened the pull request:
Update the service level cache in the node startup sequence, after the service level and auth service are initialized.

The cache update depends on the service level data accessor being set and the auth service being initialized. Before the commit, it may happen that a cache update is not triggered after the initialization. The commit adds an explicit call to update the cache where it is guaranteed to be ready.

Fixes scylladb/scylladb#21763

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13582/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13582/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
#### Failed Tests (30/150):
- [test_service_level_cache_after_restart.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.1)
- [test_service_level_cache_after_restart.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.10)
- [test_service_level_cache_after_restart.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.2)
- [test_service_level_cache_after_restart.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.3)
- [test_service_level_cache_after_restart.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.4)
- [test_service_level_cache_after_restart.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.5)
- [test_service_level_cache_after_restart.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.6)
- [test_service_level_cache_after_restart.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.7)
- [test_service_level_cache_after_restart.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.8)
- [test_service_level_cache_after_restart.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13582/testReport/junit/auth_cluster/test_raft_service_levels/test_service_level_cache_after_restart_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_level_cache_after_restart.debug.9)
#### Build Details:
- Duration: 26 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13585/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13585/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/498/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2026/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5136/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13151/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13585/)
#### Failed Tests (643/37780):
- [test_compact_tombstones_when_memtable_flush_one_node_stopped](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest___test_compact_tombstones_when_memtable_flush_one_node_stopped) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compact_tombstones_when_memtable_flush_one_node_stopped)
- [test_dynamo_reads_after_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest___test_dynamo_reads_after_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dynamo_reads_after_repair)
- [test_short_read_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_short_read_delete) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_delete)
- [test_compact_tombstones_when_memtable_flush_one_node_stopped](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest___test_compact_tombstones_when_memtable_flush_one_node_stopped_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compact_tombstones_when_memtable_flush_one_node_stopped)
- [test_dynamo_reads_after_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest___test_dynamo_reads_after_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_dynamo_reads_after_repair)
- [test_short_read_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_short_read_delete_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_delete)
- [test_data_query_diges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_data_query_diges) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_query_diges)
- [test_data_query_diges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_data_query_diges_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_query_diges)
- [test_incomplete_result_test_partition_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_incomplete_result_test_partition_limit) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_partition_limit)
- [test_incomplete_result_test_partition_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13585/testReport/junit/consistency_test/TestConsistency/Tests___dtest___test_incomplete_result_test_partition_limit_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_partition_limit)
#### Build Details:
- Duration: 4 hr 33 min
- Builder: spider6.cloudius-systems.com

mlitvk commented:
looking into an issue...

Jadw1 reviewed (approved):
looks good

Jadw1 commented on main.cc:
```diff
@@ -1979,6 +1979,9 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             });
 
+            // update the service level cache after the SL data accessor and auth service are initialized.
+            sl_controller.local().update_cache(qos::update_both_cache_levels::yes).get();
```
Oh actually, we don't want to do it unconditional. This should be triggered only if raft topology is enabled and service levels were migrated to v2.

I think that's why dtests failed but dtests with topology changes and with tablets passed. (I've only took a quick look at one random dtest)

mlitvk replied:
I changed it to trigger only in v2, thanks.

In the issue I encountered I think what happens is that the legacy loop does an access to the topology state machine, and in the shutdown sequence the tsm is destroyed before the loop is stopped. Maybe it was exposed by some change of timing.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13627/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13627/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/522/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5171/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13187/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2062/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13627/)
#### Build Details:
- Duration: 5 hr 26 min
- Builder: i-0de2816920720126d (m5d.12xlarge)

mlitvk commented:
* rebased the branch
* I added in 3 new commits some small changes to utilize the service level cache in LIST SL, ATTACH, ALTER statements and `attributes::get_service_level` instead of reading from the table

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13678/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13678/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/560/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2115/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13239/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5223/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13678/)
#### Build Details:
- Duration: 4 hr 23 min
- Builder: spider1.cloudius-systems.com

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -259,13 +259,19 @@ public:
      * get_service_level("default") otherwise.
      */
-    service_level& get_service_level(sstring service_level_name) {
+    service_level& get_service_level(sstring service_level_name, bool or_default) {
```
Nit: Personally, I think it would be a bit more clear to have two separate functions rather than having one function controller by a boolean parameter. So: `get_service_level` and `get_service_level_or_default`, for example.

mlitvk replied:
done

mlitvk commented on service/qos/qos_common.hh:
```diff
@@ -90,7 +90,7 @@ public:
 public:
     nonexistant_service_level_exception(sstring service_level_name)
-            : service_level_argument_exception(format("Service Level {} doesn't exists.", service_level_name)) {
+            : service_level_argument_exception(format("Service Level {} doesn't exist.", service_level_name)) {
```
done, I put it in a separate commit, and additionally removed an unused exception

mlitvk commented:
* instead of `get_service_level` with `or_default` parameter, now there are two functions `get_service_level` and `get_service_level_or_default`
* separate commit for small SL exception code cleanup, and removed an unused exception class

avikivity commented:
> backport needed because it's a bug fix, but maybe I should take only the first commit for the backport

If that's the case then file the first patch separately, and follow up with the others later.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13715/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13715/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/591/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5256/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13271/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2153/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13715/)
#### Build Details:
- Duration: 4 hr 34 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13735/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13735/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_raft_service_levels
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/607/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13297/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2183/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13735/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5305/)
#### Build Details:
- Duration: 5 hr 27 min
- Builder: spider6.cloudius-systems.com

piodul commented:
Queued
